plugins/dagsvisualizer/frontend/src/graph/cytoscape.ts

```diff
@@ -137,34 +140,61 @@ export function drawTransaction(
                 source: inputNodeID,
                 target: outputNodeID
             },
-            classes: 'invisible'
+            classes: 'invisible',
         })
     );
 
     graph.layoutApi.placeNewNodes(collection);
 }
 
-export function drawBranch(
+export async function drawBranch(
     branch: branchVertex,
     graph: cytoscapeLib,
     branchMap: ObservableMap<string, branchVertex>
 ) {
-    const v = graph.cy.add({
-        group: 'nodes',
-        data: { id: branch.ID }
-    });
+    if (!branch) {
+        return;
+    }
+    let v: any;
+    try {
+        v = graph.cy.add({
+            group: 'nodes',
+            data: { id: branch.ID },
+        });
+    } catch (e) {
+        // already exists. never mind
+    } finally {
+        branchMap.set(branch.ID, branch);
+    }
 
-    branch.parents.forEach(pID => {
+    branch.parents = branch.parents || [];
+    for (let i = 0; i < branch.parents.length; i++) {
+        const pID = branch.parents[i];
         const b = branchMap.get(pID);
         if (b) {
             graph.cy.add({
                 group: 'edges',
-                data: { source: pID, target: branch.ID }
+                data: { source: pID, target: branch.ID },
             });
+        } else {
+            // recursively fetch branch and draw parent
+            const res = await fetch(`/api/dagsvisualizer/branch/${pID}`);
+            const vertex: branchVertex = (await res.json()) as branchVertex;
+            console.log('parent found: ', vertex);
+            await drawBranch(vertex, graph, branchMap);
+            // make sure the parent was added
+            if (branchMap.get(pID)) {
+                graph.cy.add({
+                    group: 'edges',
+                    data: { source: pID, target: branch.ID },
+                });
+            }
```

Comment:
What about the API returns a list of branches until master and we draw them at once? I'm afraid the recursive can grow too deep, also frequent calls of layoututility might slow down the browser.

Reply:
I can try this.